Enable run and debug Scala CLI scripts in Metals

lwronski opened the pull request:
This PR adds support for running and debugging Scala CLI scripts in Metals.


https://user-images.githubusercontent.com/46607934/199930822-9ca5cb23-7708-45e8-9c4d-6e4413adbe0e.mov

Commit:
use pc for local rename

Commit:
overriden symbols

tgodzik committed:
feature: Add shell command required to run main classes

This allows clients (for example VS Code) to simply run a main class within a local terminal.

This both adds the shell command to the main class lenses as well as allows the clients to discover main classes to run together with the command to use.

Commit:
fix: correctly import companion object for apply method

Commit:
backticks and named params

Commit:
apply suggestions

tgodzik committed:
chore: Add tests for running shell command from code lenses

kpodsiad committed:
remove (just temporary) support for running main classes via other ways than lenses

kpodsiad committed:
proposition how to improve run lenses

kpodsiad committed:
remove unnecessary discriminator

lwronski committed:
Enable run and debug Scala CLI scripts in Metals

lwronski committed:
Enable run and debug Scala CLI scripts in Metals

tgodzik committed:
feature: Add completions for plugins in ScalaCLI

Fixes https://github.com/scalameta/metals/issues/4594

tgodzik committed:
chore: Remove kind from scaladoc

Commit:
fix highlight for named arg

tgodzik committed:
Don't sent jvmRunEnvironment request for ScalaCLI

tgodzik committed:
bugfix: Fix wrong namePos on package objects

Previously we would be using .name on package object and we would get `package`, which is not the correct name. Now, we look for the actual name of the package that this package object is accompanying.

tgodzik committed:
bugfix: Fix issues with code lenses tests

scalameta-bot committed:
build(deps): Update flyway-core from 9.6.0 to 9.7.0

Commit:
adjust pos

Gedochao commented on metals/src/main/scala/scala/meta/internal/metals/codelenses/RunTestCodeLens.scala:
```diff
@@ -56,25 +56,33 @@ final class RunTestCodeLens(
+      // see https://github.com/build-server-protocol/build-server-protocol/pull/161
+      // most of the bsp servers such as bloop and sbt might not support it.
+      if buildTarget.getCapabilities.getCanDebug || connection.isBloop || connection.isSbt
+      if connection.isScalaCLI || !path.isAmmoniteScript || !path.isWorksheet
```
why in a separate `if` clause?

Commit:
fix: dont show incorrect extension keyword completion

tgodzik committed:
bugfix: Don't show string code actions when they are enclosed within range

Previously, string code actions would show up in force when for example selecting the entire file to import all symbols, which would cause a bit of spam on the screen. Now, we show add/remove interpolation code actions if cursor/range is within the string

tgodzik committed:
docs: Add spearate information about scripts

We added some info in the previous release notes, but it's better to keep it in stable documentation.

I think we can start moving some feature docs to features tab to be easier to discover.

ckipp01 committed:
chore(deps): bump millw to 0.4.4

tgodzik committed:
bugfix: Default to empty jvm environment instead of failing

tgodzik committed:
bugfix: Fix compilation issues with Scala 3.3.x

Commit:
fix for interpolator completions

ckipp01 committed:
chore(ci): remove Steward run from this repo (#4613)

Instead of running Steward from this repo we now have this setup org
wide at https://github.com/scalameta/steward.

ckipp01 committed:
chore: swap scalameta Mill g8 for com-lihaoyi one (#4614)

There is an official one now support in the com-lihaoyi org. I think we
should switch out ours for this one. Then we can actually just archive
the one we have with a message to use the com-lihaoyi instead. Less
repos to maintain ha.

tgodzik committed:
bugfix: Try to make DAP suites less flaky

Commit:
add showing semanticdb for sbt files and worksheets

shardulc committed:
Add lsp-keep-workspace-alive to emacs config

shardulc committed:
Explain lsp-keep-workspace-alive

tgodzik committed:
bugfix: Retry flaky file watcher suite

tgodzik committed:
chore: Add support for Scala 3.2.2-RC1

tgodzik committed:
bugfix: Correctly display statuses for Ammonite scripts in Doctor

Previously, we would report issues with Java support and semanticdb, which isn't true, since semanticdb is generated and Java is not supported. Now, we correctly show that in the doctor.

Fixes https://github.com/scalameta/metals/issues/3305

tgodzik committed:
bugfix: Fix issues with problem resolver suite

tgodzik committed:
chore: Update nightly release branch to use Scala 3.3.x fixes

Commit:
fix: complete already overriden symbols

Commit:
fix: more accurate signature on error in override comp

Commit:
add showing completions for incomplete symbol

Commit:
move keyword completions

tgodzik committed:
bugfix: Make scripts work even if not at the top directory

tgodzik committed:
bugfix: Make scripts work even if not at the top directory

adpi2 committed:
Update scala-debug-adapter to 3.0.4

adpi2 committed:
Drop support for scala-debug-adapter 1.x

This will only affects users of Bloop < 1.4.10.
Bloop 1.4.10 was released on Oct 2021

Commit:
fix: implementAll with type aliases and higher kind type

scalameta-bot committed:
build(deps): Update interface from 1.0.10 to 1.0.11 (#4633)

scalameta-bot committed:
build(deps): Update mill-contrib-testng from 0.10.8 to 0.10.9 (#4631)

ckipp01 committed:
docs: remove old resolver in docs (#4627)

This is no longer used and doesn't need to be listed
in the bootstrap instructions

scalameta-bot committed:
build(deps): Update flyway-core from 9.7.0 to 9.8.1

scalameta-bot committed:
build(deps): Update cli_3, scala-cli-bsp from 0.1.16 to 0.1.17

scalameta-bot committed:
build(deps): Update sbt, scripted-plugin from 1.7.3 to 1.8.0 (#4635)

scalameta-bot committed:
build(deps): Update org.eclipse.lsp4j, ... from 0.17.0 to 0.18.0 (#4634)

Commit:
fix: match completion with -no-indent option

tgodzik committed:
bugfix: Fix compliation issues with 3.3.x

Previously, toTextRefinement would return `Closed` type and now it returns `Text`, which is causing the compilation issues.

Now, since the printer prints the types for refinments correctly, we can avoid overriding the method altogether.

The type is printer a bit differently, but it does compile, so let's keep it in line with the compiler.

scalameta-bot committed:
build(deps): Update coursier from 2.1.0-M7 to 2.1.0-RC1 (#4632)

* build(deps): Update coursier from 2.1.0-M7 to 2.1.0-RC1

* Revert commit(s) 1c48a6e6f0

* build(deps): Update coursier from 2.1.0-M7 to 2.1.0-RC1

kpodsiad committed:
fix: prevent NPE in import missing symbol code action (#4645)

ckipp01 committed:
chore(ci): remove coc-metals from issue template

It's been archived for a while now so we can remove this.

lwronski committed:
Add scala-debug-adapter 2x support by Scala CLI (#4649)

* Fix typo in error message

* Add scala-debug-adapter 2x supports by Scala CLI

ckipp01 committed:
refactor: ensure we're testing the connection with Mill (#4642)

* refactor: ensure we're testing the connection with Mill

When I opened https://github.com/com-lihaoyi/mill/issues/2123 it made me
wonder why we didn't hit on this in our tests, but then I realized we
were actually never testing the initialization with Mill server. This
adds some more tests to ensure we're not only testing the generation of
the BSP config file, but also connection to the build server.

* deps: update millw to 0.4.5

scalameta-bot committed:
build(deps): Update coursier from 2.1.0-RC1 to 2.1.0-RC2 (#4651)

ckipp01 committed:
fix: ensure name for scala-cli is correct (#4653)

This is a follow-up to https://github.com/scalameta/metals/pull/4649
since it didn't actually fix the issue since `isScalaCli` was looking
for the connection name to be `ScalaCli`, however it's actually
`scala-cli` when you look at the value of `name` when that check is
done. You can also see this in the `build/initialize`.

```
[Trace - 11:38:07 am] Received response 'build/initialize - (1)' in 615ms
Result: {
  "displayName": "scala-cli",
  "version": "0.1.17",
  ...
```

jkciesluk committed:
change version in ivy completions tests (#4663)

scalameta-bot committed:
build(deps): Update munit from 1.0.0-M6 to 1.0.0-M7 (#4661)

Commit:
fix: Don't show wildcard param completions

scalameta-bot committed:
build(deps): Update scribe, scribe-file, scribe-slf4j from 3.10.4 to 3.10.5 (#4658)

tgodzik committed:
bugfix: Don't add package for worksheets and scripts

scalameta-bot committed:
build(deps): Update flyway-core from 9.8.1 to 9.8.2 (#4660)

tgodzik committed:
chore: Update  sbt-ci-releas to new org

scalameta-bot committed:
build(deps): Update org.eclipse.lsp4j, ... from 0.18.0 to 0.19.0 (#4659)

susliko committed:
fix: implement methods with path-dependent types

scalameta-bot committed:
build(deps): Update bloop-config, bloop-launcher from 1.5.4-14-7cb43276 to 1.5.4-54-cfc03bdc

scalameta-bot committed:
build(deps): Update flyway-core from 9.8.2 to 9.8.3 (#4673)

tgodzik committed:
Change ScalaCLI to Scala CLI

Co-authored-by: LaurenceWarne <17688577+LaurenceWarne@users.noreply.github.com>
Co-authored-by: Chris Kipp <ckipp@pm.me>

tgodzik committed:
bugfix: Add test tag to projects containg test libraries

ckipp01 committed:
refactor: avoid newlines in json doctor output (#4676)

Super small change that I'm currently hitting on in Neovim. When
consuming the json output for the doctor there used to not be any
newlines in the output. Once we added the Java problems we have one
value that has multiple lines. Not all clients will be able to handle
new lines here, so this small change instead just splits on the newline
and gives them back as two items in the array. Again, this only happens
when using the json output.

scalameta-bot committed:
build(deps): Update bloop-config, bloop-launcher from 1.5.4-54-cfc03bdc to 1.5.4-67-c910a45b

lwronski committed:
Start Scala CLI server with server name - ScalaCli.name

tgodzik committed:
Merge branch 'enable-run-scala-scripts' of github.com:lwronski/metals into enable-run-scala-scripts

tgodzik committed:
bugfix: Make ScalaCLI breakpoint tests work without project name

tgodzik committed:
bugfix: Fix StepDapSuite
